peps/pep-0705.rst

```diff
@@ -310,235 +217,260 @@ This is consistent with the behavior introduced in :pep:`655`.
+
+  * the associated value type in ``C`` is of top type
+  * the key is required in ``B``
+  * the key is in both ``A`` and ``B``
+
+* if a key is in both ``A`` and ``C``, but not required in ``B``, the associated value type in ``A`` is consistent with the value type in ``C``
+* if a key is in both ``B`` and ``C``, the associated value type in ``B`` is consistent with the value type in ``C``
+* if a key is required in ``C``, it is required in either ``A`` or ``B``
 
 Notes:
 
-* The read-only status of the keys does not matter. A key can be read-only on just ``A``, just ``B``, or just ``C``, or any combination.
-* A key found on ``A`` or ``B`` may be missed off ``C`` if it allows other keys. Type-checkers may however choose to flag this edge-case with a warning or error in some circumstances, if it is found to be a source of mistakes.
+* The read-only status of the items does not matter. An item can be read-only on just ``A``, just ``B``, or just ``C``, or any combination.
+* A key found on ``A`` or ``B`` may theoretically be missed off ``C``. Type-checkers or other linters may however choose to flag this edge-case with a warning or error, if it is found to be a source of mistakes.
+* An item in ``A`` can be inconsistent with ``C`` if it is required in ``B``. For example::
+
+    class A(TypedDict):
+      x: str
+      y: int
+
+    class B(TypedDict):
+      x: int
+
+    class C(TypedDict):
+      x: int  # Note: A's definition of "x" is not consistent with this
+      y: int
+
+    a: A = { "x": "three", "y": 4 }
+    b: B = { "x": 3 }
+    c: C = a | b  # Accepted by type checker: "x" will always come from b
+
+Update method
+-------------
+
+In addition to existing type checking rules, type checkers should error if a TypedDict with a read-only item is updated with another TypedDict that declares that key::
 
-Update Operations
+    class A(TypedDict):
+        x: ReadOnly[int]
+        y: int
+
+    a1: A = { "x": 1, "y": 2 }
+    a2: A = { "x": 3, "y": 4 }
+    a1.update(a2)  # Type check error: "x" is read-only in A
+
+Unless the declared value is of bottom type::
+
+    class B(TypedDict):
+        x: NotRequired[typing.Never]
+        y: ReadOnly[int]
+
+    def update_a(a: A, b: B) -> None:
+        a.update(b)  # Accepted by type checker: "x" cannot be set on b
+
+copy and deepcopy
 -----------------
 
-Previously, ``clear()`` and ``popitem()`` were rejected by type checkers on TypedDict objects, as they could remove required keys, some of which may not be directly visible because of structural subtyping. However, these methods should be allowed on TypedDicts objects with all keys non-read-only and non-required and with no other keys allowed::
+As with the merge operation, the result of applying the ``copy`` method, ``copy.copy`` or ``copy.deepcopy`` to a ``TypedDict`` should be consistent with any type that can hold the possible items, not just types the operand is consistent with. For example::
```

Comment:
Here again, I think this specification is backward. The spec should indicate how the expression `copy(a)` should be evaluated by a type checker. Normal type consistency rules should then apply to the resulting type.

```python
from typing import TypedDict
from typing_extensions import ReadOnly
from copy import copy

class A(TypedDict):
    x: ReadOnly[int]

class B(TypedDict):
    x: int | str

a: A = {"x": 3}
x = copy(a)

# What type should be revealed here?
reveal_type(x)

# Normal type consistency rules will determine
# whether x is compatible with B.
b: B = x
```

Reply:
Agree, this section feels too prescriptive about the exact behavior of type checkers.

Reply:
A full discussion is on the other comment -- https://github.com/python/peps/pull/3504#discussion_r1367765069 -- I didn't want to duplicate it